lib/yargs-parser-types.ts

```diff
@@ -0,0 +1,160 @@
+  newAliases: Dictionary<boolean>;
+  /** Any new argument created by opts.default, no aliases included. */
+  defaulted: Dictionary<boolean>;
+  /** The configuration loaded from the yargs stanza in package.json. */
+  configuration: Configuration;
+}
+
+export interface Configuration {
+  /** Should variables prefixed with --no be treated as negations? Default is `true` */
+  'boolean-negation': boolean;
+  /** Should hyphenated arguments be expanded into camel-case aliases? Default is `true` */
+  'camel-case-expansion': boolean;
+  /** Should arrays be combined when provided by both command line arguments and a configuration file? Default is `false`  */
+  'combine-arrays': boolean;
+  /** Should keys that contain `.` be treated as objects? Default is `true` */
+  'dot-notation': boolean;
+  /** Should arguments be coerced into an array when duplicated? Default is `true` */
+  'duplicate-arguments-array': boolean;
+  /** Should array arguments be coerced into a single array when duplicated? Default is `true` */
+  'flatten-duplicate-arrays': boolean;
+  /** Should arrays consume more than one positional argument following their flag? Default is `true` */
+  'greedy-arrays': boolean;
+  /** Should parsing stop at the first text argument? This is similar to how e.g. ssh parses its command line. Default is `false` */
+  'halt-at-non-option': boolean;
+  /** Should nargs consume dash options as well as positional arguments? Default is `false` */
+  'nargs-eats-options': boolean;
+  /** The prefix to use for negated boolean variables. Default is `'no-'` */
+  'negation-prefix': string;
+  /** Should keys that look like numbers be treated as such? Default is `true` */
+  'parse-numbers': boolean;
+  /** Should unparsed flags be stored in -- or _? Default is `false` */
+  'populate--': boolean;
+  /** Should a placeholder be added for keys not set via the corresponding CLI argument? Default is `false` */
+  'set-placeholder-key': boolean;
+  /** Should a group of short-options be treated as boolean flags? Default is `true` */
+  'short-option-groups': boolean;
+  /** Should aliases be removed before returning results? Default is `false` */
+  'strip-aliased': boolean;
+  /** Should dashed keys be removed before returning results? This option has no effect if camel-case-expansion is disabled. Default is `false` */
+  'strip-dashed': boolean;
+  /** Should unknown options be treated like regular arguments? An unknown option is one that is not configured in opts. Default is `false` */
+  'unknown-options-as-args': boolean;
+}
+
+export type ArrayOption = string | { key: string; boolean?: boolean, string?: boolean, number?: boolean, integer?: boolean };
+
+export type CoerceCallback = (arg: any) => any;
+
+export type ConfigCallback = (configPath: string) => { [key: string]: any } | Error;
+
+export interface Options {
+  /** An object representing the set of aliases for a key: `{ alias: { foo: ['f']} }`. */
+  alias?: Dictionary<string | string[]>;
```

Comment:
Options properties should not be all optional, for a better reuse in `yargs`.

You should use instead `Partial<Options>` when you need to use all optional options.

Reply:
@mleguen I'll go ahead and make this change but I'm not sure I agree with it. Perhaps you could expand on how this makes it better for reuse in `yargs`?

While both methods are practically the same, in my option it is a better practice to explicitly set the properties as optional for this use case. None of the options are required and they all can indeed be undefined. I think using Partial is fine when reusing an existing type and/or there are clear default values, such as with Configuration, but Options doesn't really have clear defaults for all its properties. A quick test with using Partial on the options parse argument but not with the scoped opts variable (like configuration) and setting the option properties to their fallback values (empty objects/arrays/etc...) broke a lot of tests. So we would have to use partial again there or do quite a bit of refactoring in elsewhere. From the perspective of using `yargs-parser` as a dependency I think the optional properties is clearer on how to interact with the module as well. Also if later an option is required, then all of the `Partial<Options>` uses will have to be refactored to use explicit optional properties for Options, so the explicit optional properties provides better long term support.

Reply:
> Perhaps you could expand on how this makes it better for reuse in yargs

I find it always easier to add `undefined` to a type (using `?` or `Partial`) than to remove it.

Yargs builds an internal options structure (including yargs-parser options, but not only), none of it being optional as it deals with defaults by itself. Reusing your `Options` type would have helped not writing it twice.

But if it too much of a hassle to do so, no problem for me.